presto-main/src/main/java/com/facebook/presto/cost/HistogramCalculator.java

```diff
@@ -0,0 +1,167 @@
+            }
+            return Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+        }
+
+        // we know the bounds are both known, so calculate the percentile for each bound
+        // The inclusivity arguments can be derived from the open-ness of the interval we're
+        // calculating the filter factor for
+        // e.g. given a variable with values in [0, 10] to calculate the filter of
+        // [1, 9) (openness: false, true) we need the percentile from
+        // [0.0 to 1.0) (openness: false, true) and from [0.0, 9.0) (openness: false, true)
+        // thus for the "lowPercentile" calculation we should pass "false" to be non-inclusive
+        // (same as openness) however, on the high-end we want the inclusivity to be the opposite
+        // of the openness since if it's open, we _don't_ want to include the bound.
+        Estimate lowPercentile = histogram.cumulativeProbability(range.getLow(), openLow);
+        Estimate highPercentile = histogram.cumulativeProbability(range.getHigh(), !openHigh);
+
+        // both bounds are probably infinity, use the infinite-infinite heuristic
+        if (lowPercentile.isUnknown() || highPercentile.isUnknown()) {
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+            // in the case the histogram has no values
+            if (totalDistinctValues.equals(Estimate.zero()) || range.getDistinctValuesCount() == 0.0) {
+                return Estimate.of(0.0);
+            }
+
+            // in the case only one is unknown
+            if (((lowPercentile.isUnknown() && !highPercentile.isUnknown()) ||
+                    (!lowPercentile.isUnknown() && highPercentile.isUnknown())) &&
+                    isFinite(range.length())) {
+                return useHeuristics ? Estimate.of(StatisticRange.INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR) : Estimate.unknown();
+            }
+
+            if (range.length() == 0.0) {
+                return totalDistinctValues.map(distinct -> 1.0 / distinct);
+            }
+
+            if (!isNaN(range.getDistinctValuesCount())) {
+                return totalDistinctValues.map(distinct -> min(1.0, range.getDistinctValuesCount() / distinct));
+            }
+
+            return Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+        }
+
+        // in the case the range is a single value, this can occur if the input
+        // filter range is a single value (low == high) OR in the case that the
+        // bounds of the filter or this histogram are infinite.
+        // in the case of infinite bounds, we should return an estimate that
+        // correlates to the overlapping distinct values.
+        if (lowPercentile.equals(highPercentile)) {
+            if (!useHeuristics) {
+                return Estimate.zero();
+            }
```

Comment:
Should we return a `Estimate.unknown()` or a `Estimate.zero()` here? As the comment mentioned, there may exists cases that the bounds are infinite. I'm not quite so sure, what's your thought about it?

Reply:
This case is required to return zero, or else it has adverse affects on plans for queries without statistics

Reply:
OK, got it.